docs: fastapi migration (#3985)
litestar-org/litestar: docs/migration/fastapi.rst

    * :doc:`/usage/dependency-injection`

Lifespan
~~~~~~~~

If you're using an async context manager and pass parameters to it, most likely the order of parameters is inversed between FastAPI and Litestar.

.. tab-set::

    .. tab-item:: FastAPI
        :sync: fastapi

        .. code-block:: python

            @asynccontextmanager
            async def lifespan(
                _app: FastAPI,
                app_settings: AppSettings,
            ):
                # Setup code here
                yield
                # Teardown code here

    .. tab-item:: Litestar
        :sync: litestar

        .. code-block:: python

            @asynccontextmanager
            async def lifespan(
                app_settings: AppSettings,
                _app: Litestar,
            ):
                # Setup code here
                yield
                # Teardown code here

> ~~~~~~~
> Cookies
> ~~~~~~~

Reviewer: ```suggestion

Cookies
~~~~~~~
```
is this meant to be a sub-section of lifespan

Author: nope it isnt, if I commit you're suggestion I'm fine ?

Reviewer: Yeah I think that would be good to make it inline with all other sections

> While with FastAPI you usually set cookies on the response `Response` object, in Litestar it will be set at the decorator level using the `response_cookies` keyword argument.

Reviewer: ```suggestion
While with FastAPI you usually set cookies on the response ``Response`` object, in Litestar it will be set at the decorator level using the ``response_cookies`` keyword argument.
```

Do you intend for these to be stylized just in italics or like <kbd>response_cookies</kbd>?

.. seealso::

    You can also set cookies in the handler itself also

    * :ref:`Setting Cookies dynamically <usage/responses:setting cookies dynamically>`

.. tab-set::

    .. tab-item:: FastAPI
        :sync: fastapi

        .. code-block:: python

            @app.get("/")
            async def index(response: Response) -> dict[str, str]:
                response.set_cookie(key="my_cookie", value="cookie_value")
                ...

    .. tab-item:: Litestar
        :sync: litestar

        .. code-block:: python

            @get(response_cookies={"my-cookie": "cookie-value"})
            async def handler() -> str:
                ...

Dependencies parameters
~~~~~~~~~~~~~~~~~~~~~~~
The way dependencies parameters are passed differs between FastAPI and Litestar, note the `state: State` parameter in the Litestar example.

> You can get the state either with the state kwarg in the handler or request.state but those are different from the app.state, in fact request.state just inherits from app.state.

Reviewer: ```suggestion
You can get the state either with the state kwarg in the handler or ``request.state`` but those are different from the ``app.state``, in fact `request.state` just inherits from `app.state`.
```

.. tab-set::

    .. tab-item:: FastAPI
        :sync: fastapi

        .. code-block:: python

            from fastapi import Request

            async def get_arqredis(request: Request) -> ArqRedis:
                return request.state.arqredis

    .. tab-item:: Litestar
        :sync: litestar

        .. code-block:: python

            from litestar import State

            async def get_arqredis(state: State) -> ArqRedis:
                return state.arqredis

Post json

> ~~~~~~~~~
> In FastAPI, you pass the JSON object directly as a parameter to the endpoint, which will then be validated by Pydantic. In Litestar, you use the `data` keyword argument. The data will be parsed and validated by the associated modelling library.

Reviewer: ```suggestion
~~~~~~~~~

In FastAPI, you pass the JSON object directly as a parameter to the endpoint, which will then be validated by Pydantic. In Litestar, you use the `data` keyword argument. The data will be parsed and validated by the associated modelling library.
```

.. tab-set::

    .. tab-item:: FastAPI
        :sync: fastapi

        .. code-block:: python

            class ObjectType(BaseModel):
                name: str

            @app.post("/items/")
            async def create_item(object_name: ObjectType) -> dict[str, str]:
                return {"name": object_name.name}

    .. tab-item:: Litestar
        :sync: litestar

        .. code-block:: python

            from litestar import Litestar, post
            from pydantic import BaseModel

            class ObjectType(BaseModel):
                name: str

            @post("/items/")
            async def create_item(data: ObjectType) -> dict[str, str]:
                return {"name": data.name}

Default status codes
~~~~~~~~~~~~~~~~~~~~

Post defaults to 200 in FastApi and 201 in Litestar.

Templates

> ~~~~~~~~~
> In FastAPI, you use `TemplateResponse` to render templates. In Litestar, you use the `Template` class.

Reviewer: ```suggestion
~~~~~~~~~

In FastAPI, you use `TemplateResponse` to render templates. In Litestar, you use the `Template` class.
```

Also FastAPI let you pass a dictionary while in Litestar you need to explicitly pass the context kwarg.

.. tab-set::

    .. tab-item:: FastAPI
        :sync: fastapi

        .. code-block:: python

            @app.get("/uploads")
            async def get_uploads(request: Request):
                return templates.TemplateResponse(
                    "uploads.html", {"request": request, "debug": app.state.debug}
                )

    .. tab-item:: Litestar
        :sync: litestar

        .. code-block:: python

            @get("/uploads")
            async def get_uploads(app_settings) -> Template:
                return Template(
                    name="uploads.html", context={"debug": app_settings.debug}
                )

Default handler names

> ~~~~~~~~~~~~~~~~~~~~~~~
> In FastAPI, the handler name defaults to the local name of the function. In Litestar, you need to explicitly declare the `name` parameter in the route decorator. This is important when using e.g. `url_for`.

Reviewer: ```suggestion
~~~~~~~~~~~~~~~~~~~~~~~

In FastAPI, the handler name defaults to the local name of the function. In Litestar, you need to explicitly declare the `name` parameter in the route decorator. This is important when using e.g. `url_for`.
```

.. tab-set::

    .. tab-item:: FastAPI
        :sync: fastapi

        .. code-block:: python

            @app.get("/blabla")
            async def blabla() -> str:
                return "Blabla"
        .. code-block:: html

            <a href="{{ url_for('blabla') }}">Blabla</a>

    .. tab-item:: Litestar
        :sync: litestar

        .. code-block:: python

            @get(path="/blabla", name="blabla")
            async def blabla() -> str:
                return "Blabla"

        .. code-block:: html

            <a href="{{ url_for('blabla') }}">Blabla</a>

Uploads

> ~~~~~~~
> In FastAPI, you use the `File` class to handle file uploads. In Litestar, you use the `data` keyword argument with `Body` and specify the `media_type` as `RequestEncodingType.MULTI_PART`.

Reviewer: ```suggestion
~~~~~~~

In FastAPI, you use the `File` class to handle file uploads. In Litestar, you use the `data` keyword argument with `Body` and specify the `media_type` as `RequestEncodingType.MULTI_PART`.
```

While this is more verbose, it's also more explicit and communicates the intent more clearly.

.. tab-set::

    .. tab-item:: FastAPI
        :sync: fastapi

        .. code-block:: python

            @app.post("/upload/")
            async def upload_file(files: list[UploadFile] = File(...)) -> dict[str, str]:
                return {"file_names": [file.filename for file in files]}

    .. tab-item:: Litestar
        :sync: litestar

        .. code-block:: python

            @post("/upload/")
            async def upload_file(data: Annotated[list[UploadFile], Body(media_type=RequestEncodingType.MULTI_PART)]) -> dict[str, str]:
                return {"file_names": [file.filename for file in data]}

            app = Litestar([upload_file])

Exceptions signature

> ~~~~~~~~~~~~~~~~~~~~
> In FastAPI, status code and exception details can be passed to `HTTPException` as positional arguments, while in Litestar they are set with keywords arguments, e.g. `status_code`. Positional arguments to `HTTPException` in Litestar will be added to the exception detail.

Reviewer: ```suggestion
~~~~~~~~~~~~~~~~~~~~

In FastAPI, status code and exception details can be passed to `HTTPException` as positional arguments, while in Litestar they are set with keywords arguments, e.g. `status_code`. Positional arguments to `HTTPException` in Litestar will be added to the exception detail.
```

    * :doc:`/usage/dependency-injection`

Lifespan
~~~~~~~~

If you're using an async context manager and pass parameters to it, most likely the order of parameters is inversed between FastAPI and Litestar.

.. tab-set::

    .. tab-item:: FastAPI
        :sync: fastapi

        .. code-block:: python

            @asynccontextmanager
            async def lifespan(
                _app: FastAPI,
                app_settings: AppSettings,
            ):
                # Setup code here
                yield
                # Teardown code here

    .. tab-item:: Litestar
        :sync: litestar

        .. code-block:: python

            @asynccontextmanager
            async def lifespan(
                app_settings: AppSettings,
                _app: Litestar,
            ):
                # Setup code here
                yield
                # Teardown code here

~~~~~~~
Cookies
~~~~~~~

> While with FastAPI you usually set cookies on the response ``Response`` object, in Litestar it will be set at the decorator level using the ``response_cookies`` keyword argument.
> 
> .. seealso::
> 
>     You can also set cookies in the handler itself also
> 
>     * :ref:`Setting Cookies dynamically <usage/responses:setting cookies dynamically>`

Reviewer 2: ```suggestion
While with FastAPI you usually set cookies on the response ``Response`` object, in Litestar there are two options: At the decorator level, using the ``response_cookies`` keyword argument, or dynamically at the response level (see: :ref:`Setting Cookies dynamically <usage/responses:setting cookies dynamically>)
```

.. tab-set::

    .. tab-item:: FastAPI
        :sync: fastapi

        .. code-block:: python

            @app.get("/")
            async def index(response: Response) -> dict[str, str]:
                response.set_cookie(key="my_cookie", value="cookie_value")
                ...

    .. tab-item:: Litestar
        :sync: litestar

        .. code-block:: python

            @get(response_cookies={"my-cookie": "cookie-value"})
            async def handler() -> str:
                ...

Dependencies parameters
~~~~~~~~~~~~~~~~~~~~~~~
The way dependencies parameters are passed differs between FastAPI and Litestar, note the `state: State` parameter in the Litestar example.

> You can get the state either with the state kwarg in the handler or ``request.state`` but those are different from the ``app.state``, in fact `request.state` just inherits from `app.state`.

Reviewer 2: ```suggestion
You can get the state either with the state kwarg in the handler or ``request.state`` (which point to the same object, a request local state, inherited from the application's state), or via `request.app.state`, the application's state.
```

    * :doc:`/usage/dependency-injection`

Lifespan
~~~~~~~~

If you're using an async context manager and pass parameters to it, most likely the order of parameters is inversed between FastAPI and Litestar.

.. tab-set::

    .. tab-item:: FastAPI
        :sync: fastapi

        .. code-block:: python

            @asynccontextmanager
            async def lifespan(
                _app: FastAPI,
                app_settings: AppSettings,
            ):
                # Setup code here
                yield
                # Teardown code here

    .. tab-item:: Litestar
        :sync: litestar

        .. code-block:: python

            @asynccontextmanager
            async def lifespan(
                app_settings: AppSettings,
                _app: Litestar,
            ):
                # Setup code here
                yield
                # Teardown code here

~~~~~~~
Cookies
~~~~~~~

> While with FastAPI you usually set cookies on the response `Response` object, in Litestar it will be set at the decorator level.

Reviewer 2: Should probably add that (and how) you can set cookies on the response as well

Author: I added a link to dynamic cookies if that's ok for you

.. tab-set::

    .. tab-item:: FastAPI
        :sync: fastapi

        .. code-block:: python

            @app.get("/")
            async def index(response: Response) -> dict[str, str]:
                response.set_cookie(key="my_cookie", value="cookie_value")
                ...

    .. tab-item:: Litestar
        :sync: litestar

        .. code-block:: python

            @get(response_cookies={"my-cookie": "cookie-value"})
            async def handler() -> str:
                ...

Dependencies parameters
~~~~~~~~~~~~~~~~~~~~~~~

> The way dependencies parameters are passed differs between FastAPI and Litestar, note the `state: State` parameter in the Litestar example.

Reviewer 2: maybe point out the difference between `state`, `request.state` and `request.app.state`?

Author: well i wrote something, I find it confusing, so im hapy to a better rewording

.. tab-set::

    .. tab-item:: FastAPI
        :sync: fastapi

        .. code-block:: python

            from fastapi import Request

            async def get_arqredis(request: Request) -> ArqRedis:
                return request.state.arqredis

    .. tab-item:: Litestar
        :sync: litestar

        .. code-block:: python

            from litestar import State

            async def get_arqredis(state: State) -> ArqRedis:
                return state.arqredis

Post json
~~~~~~~~~

> In FastAPI, you pass the JSON object directly as a parameter to the endpoint, you are forced to use pydantic. In Litestar, you use the `data` keyword argument and can pass whatever library supported.

Reviewer 2: ```suggestion
In FastAPI, you pass the JSON object directly as a parameter to the endpoint, which will then be validated by Pydantic. In Litestar, you use the `data` keyword argument. The data will be parsed and validated by the associated modelling library.
```

.. tab-set::

    .. tab-item:: FastAPI
        :sync: fastapi

        .. code-block:: python

            class ObjectType(BaseModel):
                name: str

            @app.post("/items/")
            async def create_item(object_name: ObjectType) -> dict[str, str]:
                return {"name": object_name.name}

    .. tab-item:: Litestar
        :sync: litestar

        .. code-block:: python

            from litestar import Litestar, post
            from pydantic import BaseModel

            class ObjectType(BaseModel):
                name: str

            @post("/items/")
            async def create_item(data: ObjectType) -> dict[str, str]:
                return {"name": data.name}

Media type
~~~~~~~~~~

> In FastAPI, you specify the response type using `-> HTMLResponse` for instance. In Litestar, you use the `media_type` parameter in the decorator.

Reviewer 2: I think FastAPI also allows setting `media_type` on the response object?

Author: ok thinking about this now it doesnt make too much sense to mention that since it is incorrect on the fastapi side and it's pretty similar, so i deleted that

            @get(path="/page", media_type=MediaType.HTML)
            async def get_page() -> str:
                return "<html><body><h1>Hello, World!</h1></body></html>"

Default status codes
~~~~~~~~~~~~~~~~~~~~

Post defaults to 200 in FastApi and 201 in Litestar.

Templates
~~~~~~~~~
In FastAPI, you use `TemplateResponse` to render templates. In Litestar, you use the `Template` class.
Also FastAPI let you pass a dictionary while in Litestar you need to explicitly pass the context kwarg.

.. tab-set::

    .. tab-item:: FastAPI
        :sync: fastapi

        .. code-block:: python

            @app.get("/uploads")
            async def get_uploads(request: Request):
                return templates.TemplateResponse(
                    "uploads.html", {"request": request, "debug": app.state.debug}
                )

    .. tab-item:: Litestar
        :sync: litestar

        .. code-block:: python

            @get("/uploads")
            async def get_uploads(app_settings) -> Template:
                return Template(
                    name="uploads.html", context={"debug": app_settings.debug}
                )

Default handler names
~~~~~~~~~~~~~~~~~~~~~~~

> In FastAPI, the `url_for` function automatically gets the endpoint name. In Litestar, you need to explicitly declare the `name` parameter in the route decorator.

Reviewer 2: ```suggestion
In FastAPI, the handler name defaults to the local name of the function. In Litestar, you need to explicitly declare the `name` parameter in the route decorator. This is important when using e.g. `url_for`.
```

.. tab-set::

    .. tab-item:: FastAPI
        :sync: fastapi

        .. code-block:: python

            @app.get("/blabla")
            async def blabla() -> str:
                return "Blabla"
        .. code-block:: html

            <a href="{{ url_for('blabla') }}">Blabla</a>

    .. tab-item:: Litestar
        :sync: litestar

        .. code-block:: python

            @get(path="/blabla", name="blabla")
            async def blabla() -> str:
                return "Blabla"

        .. code-block:: html

            <a href="{{ url_for('blabla') }}">Blabla</a>
Uploads
~~~~~~~
In FastAPI, you use the `File` class to handle file uploads. In Litestar, you use the `data` keyword argument with `Body` and specify the `media_type` as `RequestEncodingType.MULTI_PART`.

> So Litestar is more verbose but less magic let you understand quicker what you are doing.

Reviewer 2: ```suggestion
While this is more verbose, it's also more explicit and communicates the intent more clearly.
```

.. tab-set::

    .. tab-item:: FastAPI
        :sync: fastapi

        .. code-block:: python

            @app.post("/upload/")
            async def upload_file(files: list[UploadFile] = File(...)) -> dict[str, str]:
                return {"file_names": [file.filename for file in files]}

    .. tab-item:: Litestar
        :sync: litestar

        .. code-block:: python

            @post("/upload/")
            async def upload_file(data: Annotated[list[UploadFile], Body(media_type=RequestEncodingType.MULTI_PART)]) -> dict[str, str]:
                return {"file_names": [file.filename for file in data]}

            app = Litestar([upload_file])

Exceptions signature
~~~~~~~~~~~~~~~~~~~~

> In FastAPI, you can pass arguments directly to `HTTPException`. In Litestar, you need to use keyword arguments, including `status_code`.

Reviewer 2: ```suggestion
In FastAPI, status code and exception details can be passed to `HTTPException` as positional arguments, while in Litestar they are set with keywords arguments, e.g. `status_code`. Positional arguments to `HTTPException` in Litestar will be added to the exception detail.
```
